src/main/java/io/confluent/idesidecar/restapi/cache/Clients.java

```diff
@@ -0,0 +1,129 @@
+package io.confluent.idesidecar.restapi.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.CaffeineSpec;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import io.confluent.idesidecar.restapi.connections.ConnectionState;
+import io.confluent.idesidecar.restapi.events.Lifecycle;
+import io.quarkus.logging.Log;
+import jakarta.enterprise.event.ObservesAsync;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * Utilities to obtain and cache clients for a given connection and client ID.
+ */
+public abstract class Clients<T extends AutoCloseable> {
+
+  /**
+   * Store an instance of a Caffeine cache for each connection. The cache will store
+   * clients by client ID and its policy may be configured by the {@link #getCaffeineSpec()}
+   * method.
+   */
+  private final Map<String, Cache<String, T>> clientsByIdByConnections =
+      new ConcurrentHashMap<>();
+
+  RemovalListener<String, T> removalListener = (key, value, cause) -> {
+    try {
+      if (value != null) {
+        Log.debugf("Closing client %s", value);
+        value.close();
+      }
+    } catch (Throwable t) {
+      // Ignore these as we don't care
+    }
+  };
+
+  /**
+   * Get a client for the given connection and client ID. If the client does not
+   * already exist, it will be created using the provided factory.
+   *
+   * @param connectionId     the ID of the connection
+   * @param clientId         the identifier of the client
+   * @param factory          the method that will create the client if there is not already one
+   * @return the client
+   */
+  public T getClient(
+      String connectionId,
+      String clientId,
+      Supplier<T> factory
+  ) {
+    return clientsForConnection(connectionId).asMap().computeIfAbsent(
+        clientId,
+        k -> factory.get()
+    );
+  }
+
+  private Cache<String, T> clientsForConnection(String connectionId) {
+    return clientsByIdByConnections.computeIfAbsent(connectionId, k -> createCache());
+  }
+
+  int clientCount() {
+    return clientsByIdByConnections
+        .values()
+        .stream()
+        .map(Cache::asMap)
+        .map(Map::size)
+        .mapToInt(Integer::intValue)
+        .sum();
+  }
```

Comment:
Does this method return a primitive type because of the limitations of `IntStream.sum()`?

Reply:
Hmm, I'm not sure. FWIW I didn't author this function per se, it was moved over from `SchemaRegistryClients`. It's only used in tests.

Also, TIL about `IntStream`.